segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -596,19 +601,21 @@ public void validateAndSaveSnapshot(MetadataTransaction txn) throws Exception {
      * @param bytes Bytes to write.
      */
     private void writeToJournal(ByteArraySegment bytes) throws ExecutionException, InterruptedException {
-        val bytesWritten = chunkStorage.write(currentHandle, systemJournalOffset, bytes.getLength(),
-                new ByteArrayInputStream(bytes.array(), bytes.arrayOffset(), bytes.getLength())).get();
-        Preconditions.checkState(bytesWritten == bytes.getLength());
-        systemJournalOffset += bytesWritten;
-    }
 
-    /**
-     * Adds a new System journal file.
-     */
-    private void startNewJournalFile() throws ExecutionException, InterruptedException {
-        currentHandle = chunkStorage.create(getSystemJournalChunkName(containerId, epoch, currentFileIndex + 1)).get();
-        currentFileIndex++;
-        systemJournalOffset = 0;
+        if (newChunkRequired) {
+            currentFileIndex++;
+            systemJournalOffset = 0;
+            currentHandle = chunkStorage.createWithContent(getSystemJournalChunkName(containerId, epoch, currentFileIndex), bytes.getLength(),
+                    new ByteArrayInputStream(bytes.array(), bytes.arrayOffset(), bytes.getLength())).get();
+            systemJournalOffset += bytes.getLength();
+            newChunkRequired = false;
+        } else {
+            Preconditions.checkState(chunkStorage.supportsAppend() && config.isAppendEnabled());
+            val bytesWritten = chunkStorage.write(currentHandle, systemJournalOffset, bytes.getLength(),
+                    new ByteArrayInputStream(bytes.array(), bytes.arrayOffset(), bytes.getLength())).get();
+            Preconditions.checkState(bytesWritten == bytes.getLength());
```

Comment:
and if this fails? How do you roll back? What debugging info do you include in the error message to quickly pinpoint what went awry?

Reply:
fixed.